src/ticks/exp_engine.c

```diff
@@ -0,0 +1,517 @@
+    to->type.type_ = TYPE_GENERIC_POINTER;
+    to->type.next = type_of;
+
+    switch (from->type.type_) {
+        case TYPE_FLOAT: {
+            to->as_pointer.ptr = (uint16_t)from->as_float;
+            to->as_pointer.element_size = from->memory_size;
+            return;
+        }
+        case TYPE_SHORT:
+        case TYPE_CHAR:
+        case TYPE_INT:
+        case TYPE_LONG: {
+            to->as_pointer.ptr = (uint16_t)from->as_int;
+            to->as_pointer.element_size = from->memory_size;
+            return;
+        }
+        default: {
+            to->flags |= EXPRESSION_ERROR;
+            char tp[128];
+            expression_result_type_to_string(&from->type, is_expression_signed(from), tp);
+            sprintf(to->as_error, "Cannot convert type %s to a pointer", tp);
+            return;
+        }
+    }
+}
+
+void expression_dereference_pointer(struct expression_result_t *from, struct expression_result_t *to) {
+    extern backend_t bk;
+    if (!(is_expression_a_pointer(from))) {
+        to->flags = EXPRESSION_ERROR;
+        char tp[128];
+        expression_result_type_to_string(&from->type, is_expression_signed(from), tp);
+        sprintf(to->as_error, "Cannot dereference type: %s", tp);
+        return;
+    }
+
+    if (from->type.next == NULL) {
+        to->flags |= EXPRESSION_ERROR;
+        sprintf(to->as_error, "Cannot dereference void");
+        return;
+    }
+
+    to->type = *from->type.next;
+    to->type.next = copy_type_chain(to->type.next);
+    to->memory_location = from->as_pointer.ptr;
+    to->memory_size = from->as_pointer.element_size;
+
+    int16_t data = from->as_pointer.ptr;
+
+    switch (to->type.type_) {
+        case TYPE_VOID: {
+            to->flags |= EXPRESSION_ERROR;
+            sprintf(to->as_error, "Cannot dereference void");
+            return;
+        }
+        case TYPE_CHAR: {
+            to->as_int = bk.get_memory(data);
+            break;
+        }
+        case TYPE_INT: {
+            to->as_int = (bk.get_memory(data + 1) << 8) + bk.get_memory(data);
+            break;
+        }
+        case TYPE_LONG:{
+            to->as_int = (bk.get_memory(data + 3) << 24) + (bk.get_memory(data + 2) << 16) + (bk.get_memory(data + 1) << 8) + bk.get_memory(data);
+            break;
+        }
+        case TYPE_FLOAT: {
+            to->flags |= EXPRESSION_ERROR;
+            sprintf(to->as_error, "Cannot dereference float (not implemented)");
+            break;
+        }
+        default: {
+            to->flags |= EXPRESSION_ERROR;
+            char tp[128];
+            expression_result_type_to_string(&from->type, is_expression_signed(from), tp);
+            sprintf(to->as_error, "Cannot dereference type (not implemented): %s", tp);
+            break;
+        }
+    }
+}
+
+type_chain expression_string_get_type(const char* str, uint8_t* is_signed) {
+    type_chain result = {};
+    if (strcmp(str, "void") == 0) {
```

Comment:
I think this might be better/more easily maintained as an array of structs:

struct {
   char *typename;
   char.  assigned;
   enum type_record_type;
} types[] = {
  { "void", 0, TYPE_VOID },
  ...
};

And just iterating through - will make it easier to add the 64 bit/float types

Reply:
Thanks.